### ISSUE TITLE
need contributor / plug-in guidelines on aliases and disabling setting of aliases

### ISSUE THREAD

**Issue by ncolton** [ohmyzsh/ohmyzsh#6630]
Some good discussion has already happened in #5721 but this is an issue not just limited to the plugin for git.

Some examples:
#4780
#5214
#5288
#5313
#5766
#5774
#5783
#6062
#6601

There needs to be some form of guidance on how to add aliases while being a good citizen, especially since some aliases can be very dangerous and destructive. It would be nice to adopt the added functionality that many plugins provide, but have the option to use, or not use the aliases provided by the plugin.

Similarly, the libraries also exhibit some of this. `lib/directories.zsh` creates 23 aliases on it's own, `lib/misc.zsh` creates 2 different aliases for `sudo`. In comparison, `lib/correction.zsh` also creates a number of aliases, but they are gated behind an option so I can easily enable or disable them.

I've avoided some plugins because of the sheer number of aliases I'd be "forced" to have from them. For the 23 aliases from `lib/directories.zsh` I resorted to creating an empty `custom/lib/directories.zsh` to disable them. I don't consider it to be realistic to have each user have to check every update to see if any new aliases were created, then add an `unalias yet_another_alias` for every new unwanted alias.

**Comment by ncolton** [comment]
As an additional note, the aliases should probably be opt-in so they don't just overwhelm new users and you are given a chance to explore them and ideally understand them a little before turning them on.

**Comment by Shea690901** [comment]
Not only the aliases should be optin but also (un)setopt's...
e.g.: I have ```alias l=less``` (added as different alias by lib/directories.zsh) and I don't like ```setopt autopushd``` (set by lib/directories.zsh)...

**Comment by crud3** [comment]
This is definitely a good idea.
I only have the git plugin activated, and that alone adds a ton of aliases. I could live with a handful, but it's over 150.

Right now I delete all of them and restore the pre-oh-my-zsh ones in lack of a better method:
```
# Saving pre-oh-my-zsh aliases
saved_aliases=$(alias -L)
source $ZSH/oh-my-zsh.sh
# Delete all aliases, then restore the old ones
unalias -a; eval $saved_aliases; unset saved_aliases
```

**Comment by lonelykid** [comment]
This is really needed. I don't want aliases that I don't know where it is defined and how to enable/disable.

**Comment by damien** [comment]
Ran into this today with aliases defined in `lib/directories.zsh`. Had to run a ZSH process in debug mode to figure out where aliases were coming from. I have some thoughts on how opt-in aliases could be managed. Here are some thoughts:

### Current problems

1. ZSH itself and ZSH plugins define aliases in a global context.
2. Unless specifically coded into ZSH or ZSH plugins, there is no way to opt-out or configure alias definitions
3. A user has no straightforward way of knowing where an alias comes from
4. For aliases offered by plugins that are opt-in, a user has no way of knowing what aliases are available

### Proposal

Provide a mechanism within ZSH for registration of aliases. This mechanism would keep track of a mapping of aliases, their recommended labels, and the shell snippets they contain. ZSH could them provide further mechanisms to allow interaction with this alias store to do things like enable, disable, or rename an alias.

In abstract, I think it'd look something like this:

Alias Source      | Alias Label | Alias Value      | Active    | User Label
----------------- | ----------- | ---------------- | --------- | ----------
`directories.zsh` | `lsa`       | `ls -lah`        | `TRUE`    | `NULL`
`directories.zsh` | `l`         | `ls -lah`        | `FALSE`   | `NULL`
`custom/exa.zsh`  | `l`         | `exa -FG`        | `TRUE`    | `NULL`
`custom/exa.zsh`  | `ll`        | `exa --extended` | `TRUE`    | `el`

This would ideally result in the following behavior:

1. `lsa` executes `ls -lah`; the user didn't modify anything and has that alias enabled from `directories.zsh`
2. `l` would execute `exa -FG`; this conflicts with `l` defined in `directories.zsh`, but is resolved by the latter alias being inactive/disabled
3. `el` executes `exa --extended` ; `custom/exa.zsh` originally defines the command under the alias `ll`, but the user has remapped the alias with to a new name, `el`

Does this sound like overkill? I'd be happy to contribute something like this, but there'd still be open questions about interface and migration paths. Open questions off the top of my head:

1. For any new approach, how do we get existing plugins moved over with a minimum of fuss and work?
2. What would a UI/TUI for this look like?
3. How big a problem is this, really? I know it's hit me and a few other people on this thread. Is this a big enough problem/concern to justify additional complexity this would add to an already large project like oh-my-zsh?

Cheers! 🍻

**Comment by damien** [comment]
Ran into this today with aliases defined in `lib/directories.zsh`. Had to run a ZSH process in debug mode to figure out where aliases were coming from. I have some thoughts on how opt-in aliases could be managed. Here are some thoughts:

### Current problems

1. ZSH itself and ZSH plugins define aliases in a global context.
2. Unless specifically coded into ZSH or ZSH plugins, there is no way to opt-out or configure alias definitions
3. A user has no straightforward way of knowing where an alias comes from
4. For aliases offered by plugins that are opt-in, a user has no way of knowing what aliases are available

### Proposal

Provide a mechanism within ZSH for registration of aliases. This mechanism would keep track of a mapping of aliases, their recommended labels, and the shell snippets they contain. ZSH could them provide further mechanisms to allow interaction with this alias store to do things like enable, disable, or rename an alias.

In abstract, I think it'd look something like this:

Alias Source      | Alias Label | Alias Value      | Active    | User Label
----------------- | ----------- | ---------------- | --------- | ----------
`directories.zsh` | `lsa`       | `ls -lah`        | `TRUE`    | `NULL`
`directories.zsh` | `l`         | `ls -lah`        | `FALSE`   | `NULL`
`custom/exa.zsh`  | `l`         | `exa -FG`        | `TRUE`    | `NULL`
`custom/exa.zsh`  | `ll`        | `exa --extended` | `TRUE`    | `el`

This would ideally result in the following behavior:

1. `lsa` executes `ls -lah`; the user didn't modify anything and has that alias enabled from `directories.zsh`
2. `l` would execute `exa -FG`; this conflicts with `l` defined in `directories.zsh`, but is resolved by the latter alias being inactive/disabled
3. `el` executes `exa --extended` ; `custom/exa.zsh` originally defines the command under the alias `ll`, but the user has remapped the alias with to a new name, `el`

Does this sound like overkill? I'd be happy to contribute something like this, but there'd still be open questions about interface and migration paths. Open questions off the top of my head:

1. For any new approach, how do we get existing plugins moved over with a minimum of fuss and work?
2. What would a UI/TUI for this look like?
3. How big a problem is this, really? I know it's hit me and a few other people on this thread. Is this a big enough problem/concern to justify additional complexity this would add to an already large project like oh-my-zsh?

Cheers! 🍻

**Comment by mcornella** [comment]
There's a proposal for a way to turn off aliases in https://github.com/ohmyzsh/ohmyzsh/issues/10644. Please have a look and add your feedback. Thanks!

**Comment by mcornella** [comment]
This was implemented in #11550, see https://github.com/ohmyzsh/ohmyzsh#skip-aliases for how to use it.
